File: tests/fixtures/quality/csharp/Scheduler.cs
using System;
using System.Collections.Generic;

namespace Acme.Worker;

/// <summary>Anything the scheduler can drive.</summary>
public interface IRunnable
{
    /// <summary>Begin processing.</summary>
    void Start();

    int Depth { get; }
}

/// <summary>What can go wrong while dispatching.</summary>
public enum Outcome
{
    Timeout,
    Rejected
}

/// <summary>Notified after each attempt.</summary>
public delegate void AttemptHandler(int attempt);

/// <summary>A unit of deferred work.</summary>
public record JobSpec(string Name, int Attempts);

/// <summary>Runs jobs with exponential backoff between retries.</summary>
public class Scheduler : BaseWorker, IRunnable
{
    /// <summary>Bounds how often a job is retried.</summary>
    public const int MaxAttempts = 5;

    private readonly List<JobSpec> pending = new();

    /// <summary>Raised after every attempt.</summary>
    public event AttemptHandler? Attempted;

    /// <summary>Queue depth, for backpressure.</summary>
    public int Depth => pending.Count;

    /// <summary>Drain the pending queue.</summary>
    public void Start()
    {
        foreach (var spec in pending)
        {
            Dispatch(spec, MaxAttempts);
        }
    }

    /// <summary>Run one job, retrying with exponential backoff.</summary>
    public bool Dispatch(
        JobSpec spec,
        int maxAttempts)
    {
        var attempt = 0;
        while (attempt < maxAttempts)
        {
            attempt++;
        }
        return true;
    }

    private void Reset()
    {
        pending.Clear();
    }
}

File: tests/fixtures/quality/java/Scheduler.java
package com.acme.worker;

import java.util.ArrayList;
import java.util.List;

/** Anything the scheduler can drive. */
public interface Runnable {
  /** Begin processing. */
  void start();

  int depth();
}

/** A unit of deferred work. */
record JobSpec(String name, int attempts) {}

/** What can go wrong while dispatching. */
enum Outcome {
  TIMEOUT,
  REJECTED
}

/** Runs jobs with exponential backoff between retries. */
public class Scheduler extends BaseWorker implements Runnable {
  /** Bounds how often a job is retried. */
  public static final int MAX_ATTEMPTS = 5;

  private final List<JobSpec> pending = new ArrayList<>();

  /** Drain the pending queue. */
  @Override
  public void start() {
    for (JobSpec spec : pending) {
      dispatch(spec, MAX_ATTEMPTS);
    }
  }

  /** Run one job, retrying with exponential backoff. */
  public boolean dispatch(
      JobSpec spec,
      int maxAttempts) {
    int attempt = 0;
    while (attempt < maxAttempts) {
      attempt++;
    }
    return true;
  }

  private void reset() {
    pending.clear();
  }

  @Override
  public int depth() {
    return pending.size();
  }
}
